app/slave/cluster_slave.py

```diff
@@ -173,18 +175,20 @@ def _do_build_teardown_and_reset(self, timeout=None):
         for executor in self.executors_by_id.values():
             executor.kill()
 
-        if not self._project_type:
-            return  # There is no build to tear down! (Slave is idle.)
+        # This lock keeps multiple teardown attempts from racing each other and throwing exceptions.
+        with self._build_teardown_lock:
```

Comment:
This feels somehow redundant with the self._build_teardown_coin. @josephharrington ?